src/opm/output/eclipse/Tables.cpp

```diff
@@ -1006,6 +1085,876 @@ namespace { namespace SatFunc {
     } // Water
 }} // Anonymous::SatFunc
 
+/// Functions to facilitate generating TAB vector entries for tabulated
+/// PVT functions.
+namespace { namespace PVTFunc {
+
+    /// Functions to create linearised, padded, and normalised gas PVT
+    /// output tables from various input gas PVT function keywords.
+    namespace Gas {
+        /// Create linearised and padded 'TAB' vector entries of normalised
+        /// gas tables for all PVT function regions from PVDG (dry gas)
+        /// keyword data.
+        ///
+        /// \param[in] numPressNodes Number of pressure nodes (rows) to
+        ///    allocate in the output vector for each table.  Expected to be
+        ///    equal to the number of declared pressure nodes in the
+        ///    simulation run's TABDIMS keyword (NPPVT, Item 4).
+        ///
+        /// \param[in] units Active unit system.  Needed to convert SI
+        ///    convention pressure values (Pascal), formation volume factors
+        ///    (Rm3/Sm3), and viscosity values (Pascal seconds) to declared
+        ///    conventions of the run specification.
+        ///
+        /// \param[in] pvdg Collection of PVDG tables for all PVT regions.
+        ///
+        /// \return Linearised and padded 'TAB' vector values for output gas
+        ///    PVT tables.  A unit-converted copy of the input table \p pvdg
+        ///    with added derivatives.
+        std::vector<double>
+        fromPVDG(const std::size_t          numPressNodes,
+                 const Opm::UnitSystem&     units,
+                 const Opm::TableContainer& pvdg)
+        {
+            // Columns [ Pg, 1/Bg, 1/(Bg*mu_g), derivatives ]
+            using PVDG = ::Opm::PvdgTable;
+
+            const auto numTab  = pvdg.size();
+            const auto numPrim = std::size_t{1}; // No sub-tables
+            const auto numRows = numPressNodes;  // One row per pressure node
+            const auto numDep  = std::size_t{2}; // 1/Bg, 1/(Bg*mu_g)
+
+            // PVDG fill value = +2.0e20
+            const auto fillVal = +2.0e20;
+
+            return createPropfuncTable(numTab, numPrim, numRows, numDep, fillVal,
+                [&units, &pvdg](const std::size_t           tableID,
+                                const std::size_t           primID,
+                                Opm::LinearisedOutputTable& linTable)
+                -> std::size_t
+            {
+                const auto& t = pvdg.getTable<PVDG>(tableID);
+
+                auto numActRows = std::size_t{0};
+
+                // Column 0: Pg
+                {
```

Comment:
I really like this use of lexical scopes - I usually get lot's of flak for doing it, but I am happy to see others like it as well.